source/common/router/header_parser.h

```diff
@@ -14,54 +14,37 @@
 namespace Envoy {
 namespace Router {
 
-class RequestHeaderParser;
-typedef std::unique_ptr<RequestHeaderParser> RequestHeaderParserPtr;
+class HeaderParser;
+typedef std::unique_ptr<HeaderParser> HeaderParserPtr;
 
 /**
- * This class holds the parsing logic required during configuration build and
- * also perform evaluation for the variables at runtime.
+ * HeaderParser manipulates Http::HeaderMap instances. Headers to be added are pre-parsed to select
+ * between a constant value implementation and a dynamic value implementation based on
+ * AccessLog::RequestInfo fields.
  */
-class RequestHeaderParser {
+class HeaderParser {
 public:
-  virtual ~RequestHeaderParser() {}
-
-  static RequestHeaderParserPtr
-  parse(const Protobuf::RepeatedPtrField<envoy::api::v2::HeaderValueOption>& headers);
-
-  void evaluateRequestHeaders(Http::HeaderMap& headers,
-                              const AccessLog::RequestInfo& request_info) const;
-
-private:
-  std::list<std::pair<Http::LowerCaseString, HeaderFormatterPtr>> header_formatters_;
-
-  static HeaderFormatterPtr parseInternal(const std::string& format, const bool append);
-};
-
-class ResponseHeaderParser;
-typedef std::unique_ptr<ResponseHeaderParser> ResponseHeaderParserPtr;
-
-/**
- * This class holds the logic required to apply response headers additions and removals.
- */
-class ResponseHeaderParser {
-public:
-  virtual ~ResponseHeaderParser() {}
-
-  static ResponseHeaderParserPtr
-  parse(const Protobuf::RepeatedPtrField<envoy::api::v2::HeaderValueOption>& headers_to_add,
-        const Protobuf::RepeatedPtrField<ProtobufTypes::String>& headers_to_remove);
-
-  void evaluateResponseHeaders(Http::HeaderMap& headers) const;
+  /*
+   * @param headers_to_add defines the headers to add during calls to evaluateHeaders
+   * @return HeaderParserPtr a configured HeaderParserPtr
+   */
+  static HeaderParserPtr
+  configure(const Protobuf::RepeatedPtrField<envoy::api::v2::HeaderValueOption>& headers_to_add);
```

Comment:
Stylistically, I don't think I've seen the pattern of a static factory function in envoy (vs just having a public constructor).  But I don't feel strongly against it.  However, I'd say you should add "HeaderParser() = delete;" if you're going to use this pattern.

Reply:
There are some other examples of this. Upstream::ClusterImplBase is one. 

The default constructor actually gets used, so I explicitly protected it. I'm not attached to this style -- it's just what was used in the class when I started modifying it, so I kept it.